### DISCUSSION TITLE
mastodon-streaming.service not starting (fatal error) - 502 bad gateway

### DISCUSSION THREAD

**Discussion by charismeki** [mastodon/mastodon#24426] [General]
Hi,

I managed to install mastodon on Ubuntu 22 following https://www.linuxbabe.com/ubuntu/how-to-install-mastodon-on-ubuntu. The machine is a fresh vServer with DNS over cloudflare but it is not proxied. Cache Time is set to 1h. The services mastodon-web and mastodon-sidekiq are active.

This morning I updated mastodon to 4.1.1 sucessfully but the problem still persists. When posting the notification stream is not updating automatically.

<img width="1416" alt="Bildschirm­foto 2023-04-05 um 11 58 05" src="https://user-images.githubusercontent.com/1384568/230048701-61f89ebb-856d-42ab-8d07-2e1e6a6de76c.png">

I post this problem here at discussions because i do not know if it is more my own configuration problem. The site is up and running but the streaming service is crashing with the following error notification in journalctl:

Apr 05 11:01:26 post.x.de node[842169]: /var/www/html/mastodon/node_modules/jsdom/lib/api.js:139
Apr 05 11:01:26 post.x.de node[842169]:           referrer: req.getHeader("referer") ?? undefined
Apr 05 11:01:26 post.x.de node[842169]:                                               ^
Apr 05 11:01:26 post.x.de node[842169]: SyntaxError: Unexpected token '?'
Apr 05 11:01:26 post.x.de node[842169]:     at wrapSafe (internal/modules/cjs/loader.js:915:16)
Apr 05 11:01:26 post.x.de node[842169]:     at Module._compile (internal/modules/cjs/loader.js:963:27)
Apr 05 11:01:26 post.x node[842169]:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
Apr 05 11:01:26 post.x.de node[842169]:     at Module.load (internal/modules/cjs/loader.js:863:32)
Apr 05 11:01:26 post.x.de node[842169]:     at Function.Module._load (internal/modules/cjs/loader.js:708:14)
Apr 05 11:01:26 post.x.de node[842169]:     at Module.require (internal/modules/cjs/loader.js:887:19)
Apr 05 11:01:26 post.x.de node[842169]:     at require (internal/modules/cjs/helpers.js:74:18)
Apr 05 11:01:26 post.x.de node[842169]:     at Object.<anonymous> (/var/www/html/mastodon/streaming/index.js:15:19)
Apr 05 11:01:26 post.x.de node[842169]:     at Module._compile (internal/modules/cjs/loader.js:999:30)
Apr 05 11:01:26 post.x.de node[842169]:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
Apr 05 11:01:27 post.x.de systemd[1]: mastodon-streaming.service: Main process exited, code=exited, status=1/FAILURE
Apr 05 11:01:27 post.x.de systemd[1]: mastodon-streaming.service: Failed with result 'exit-code'.
Apr 05 11:01:27 post.x.de systemd[1]: mastodon-streaming.service: Scheduled restart job, restart counter is at 1091.
Apr 05 11:01:27 post.x.de systemd[1]: Stopped mastodon-streaming.

(Yes, the unit is added with Restart=on-failure but I will remove this.)
The respecting line in jsdom/lib/api.js is the same like [released.](https://github.com/jsdom/jsdom/blob/master/lib/api.js) 

--

Node version is 16.
nvm -v = v16.19.1

--

Permissions are checked, 

--

nginx access.log is runing without any errors. 
nginx error.log is showing 

`2023/04/05 11:49:14 [error] 853958#853958: *969 connect() failed (111: Unknown error) while connecting to upstream, client: 81.12.12.12, server: social.y.de, request: "GET /api/v1/streaming/? HTTP/1.1", upstream: "http://127.0.0.1:4000/api/v1/streaming/?", host: "social.y.de"`

--

 My nginx config is copied from main/dist
map $http_upgrade $connection_upgrade {
  default upgrade;
  ''      close;
}

upstream backend {
    server 127.0.0.1:3000 fail_timeout=0;
}

upstream streaming {
    server 127.0.0.1:4000 fail_timeout=0;
}

proxy_cache_path /var/cache/nginx levels=1:2 keys_zone=CACHE:10m inactive=7d max_size=1g;

server {
    if ($host = social.y.de) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

  listen 80;
  listen [::]:80;
  server_name social.y.de;
  root /var/www/html/mastodon/public;
  location /.well-known/acme-challenge/ { allow all; }
  location / { return 301 https://$host$request_uri; }

}

server {
  listen 443 ssl http2;
  listen [::]:443 ssl http2;
  server_name social.y.de;

  ssl_protocols TLSv1.2 TLSv1.3;
  ssl_ciphers HIGH:!MEDIUM:!LOW:!aNULL:!NULL:!SHA;
  ssl_prefer_server_ciphers on;
  ssl_session_cache shared:SSL:10m;
  ssl_session_tickets off;

  ssl_certificate /etc/letsencrypt/live/social.y.de/fullchain.pem; # managed by Certbot
  ssl_certificate_key /etc/letsencrypt/live/social.y.de/privkey.pem; # managed by Certbot
  keepalive_timeout    70;
  sendfile             on;
  client_max_body_size 80m;

  root /var/www/html/mastodon/public;

  gzip on;
  gzip_disable "msie6";
  gzip_vary on;
  gzip_proxied any;
  gzip_comp_level 6;
  gzip_buffers 16 8k;
  gzip_http_version 1.1;
  gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript image/svg+xml image/x-icon;

  location / {
    try_files $uri @proxy;
  }

  location = /sw.js {
    add_header Cache-Control "public, max-age=604800, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/assets/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/avatars/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/emoji/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/headers/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/packs/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/shortcuts/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/sounds/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ~ ^/system/ {
    add_header Cache-Control "public, max-age=2419200, immutable";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri =404;
  }

  location ^~ /api/v1/streaming {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";

    proxy_pass http://streaming;
    proxy_buffering off;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";

    tcp_nodelay on;
  }

  location @proxy {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";
    proxy_pass_header Server;

    proxy_pass http://backend;
    proxy_buffering on;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    proxy_cache CACHE;
    proxy_cache_valid 200 7d;
    proxy_cache_valid 410 24h;
    proxy_cache_use_stale error timeout updating http_500 http_502 http_503 http_504;
    add_header X-Cached $upstream_cache_status;

    tcp_nodelay on;
  }

  error_page 404 500 501 502 503 504 /500.html;

    ssl_trusted_certificate /etc/letsencrypt/live/social.y.de/chain.pem; # managed by Certbot
    ssl_stapling on; # managed by Certbot
    ssl_stapling_verify on; # managed by Certbot

}

#################

The .env.production is

```
LOCAL_DOMAIN=social.y.de
SINGLE_USER_MODE=false
SECRET_KEY_BASE=asdf
OTP_SECRET=fdsa
VAPID_PRIVATE_KEY=asdf
VAPID_PUBLIC_KEY=fdsa
DB_HOST=/var/run/postgresql
DB_PORT=5432
DB_NAME=mastodon
DB_USER=mastodon
DB_PASS=asdfasdfasdf
REDIS_HOST=localhost
REDIS_PORT=6379
REDIS_PASSWORD=
SMTP_SERVER=mx01.y.de
SMTP_PORT=465
SMTP_LOGIN=notifications@y.de
SMTP_PASSWORD=asdfasdfasdf
SMTP_AUTH_METHOD=plain
SMTP_SSL=true
SMTP_OPENSSL_VERIFY_MODE=none
SMTP_ENABLE_STARTTLS=auto
SMTP_DELIVERY_METHOD=smtp
SMTP_DOMAIN=localhost
SMTP_FROM_ADDRESS='y <info@y.de>'
DEFAULT_LOCALE=en
```

--

So my question is, why does the streaming service not start? Any hints appreciated.

Many thanks!

**Comment by charismeki** [comment]
Still struggeling getting the streaming service to run. Changed the nginx and mastodon-web.service to port 3001. No change. Still not starting.

How to find out if there are problems with nodejs and node, nvm and npm. Sorry for the silly question, but i am not that familiar with nodejs.
`node -v: v16.20.0`

```
nvm ls
       v14.21.3
->     v16.20.0
        v17.9.1
         system
default -> 16 (-> v16.20.0)
iojs -> N/A (default)
unstable -> N/A (default)
node -> stable (-> v17.9.1) (default)
stable -> 17.9 (-> v17.9.1) (default)
lts/* -> lts/hydrogen (-> N/A)
lts/argon -> v4.9.1 (-> N/A)
lts/boron -> v6.17.1 (-> N/A)
lts/carbon -> v8.17.0 (-> N/A)
lts/dubnium -> v10.24.1 (-> N/A)
lts/erbium -> v12.22.12 (-> N/A)
lts/fermium -> v14.21.3
lts/gallium -> v16.20.0
lts/hydrogen -> v18.15.0 (-> N/A)
```

```
npm version
{
  npm: '8.19.4',
  node: '16.20.0',
  v8: '9.4.146.26-node.26',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.19.0',
  modules: '93',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '1.1.1t+quic',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022f',
  unicode: '14.0',
  ngtcp2: '0.8.1',
  nghttp3: '0.7.0'
}
```

Any hint and help welcome to look at this problem.

**Comment by charismeki** [comment]
Nobody? I have tried really a lot. If I stop nginx the streaming service also does not start. I assume it should be starting also if nginx is stopped, correct?

**Comment by charismeki** [comment]
Is there something which I miss to configure for streaming service that is so obvious nobody wants to help ;-)

I found some related report here and there but nothing helped until now. Still any help appreciated..

**Reply by alonecomplex** [reply]
I have the same issue after upgrading to v4.1.2 (my node is v15.14.0), the unexpected token error popped up in the streaming service log. We are on this together now I guess 🥲

**Comment by charismeki** [comment]
Yes, the error also still persists in 4.1.2 for me. Please let me know if you have any further insights or a solution :-)

**Reply by alonecomplex** [reply]
Yeah totally! I was wondering did you experience any internal error 500 caused by this streaming issue? I got 500 after upgrading and trying to rule out any other factors.

**Comment by renchap** [comment]
I think this error happens because you are not running Node 16.

Your Node process chokes on the `??` syntax ([nullish coallescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing)) which is [supported in Node 15.14+](https://node.green/#ES2021-features-Logical-Assignment-----basic-support).

We strongly recommend running Node 16 or 18, as those are supported LTS versions.

I suspect you installed Node 16 using `nvm`, which detects the `.nvmrc` in Mastodon's directory and switches the version, but your systemd unit is using the system's installed Node which is older.

You need to fix this, either by ensuring your systemd unit is using `nvm` and the correct Node version, or install a more recent NodeJS version system-wide, [as we explain in the docs](https://docs.joinmastodon.org/admin/install/#node-js).

**Reply by alonecomplex** [reply]
That helped, thank you so much! 
(You guys might want to change the recommended Node to 16 in the release note - it still says 14 or higher :-)

**Reply by charismeki** [reply]
@renchap what other reasons might there be? I have node -v: v16.20.0 running on my machine.

**Reply by renchap** [reply]
@charismeki What does `which node` show?

I am pretty sure this comes from your systemd unit not using the node binary you expect, and starting the streaming server with an old node version.

**Reply by charismeki** [reply]
```
root@post:~# which node 
/root/.nvm/versions/node/v16.19.1/bin/node
```

And

```
root@post:~# sudo -u mastodon which node 
/usr/local/bin/node
```

**Reply by charismeki** [reply]
@renchap is this the expected info? It differs from node -v which gives back v16.20.0...

**Reply by renchap** [reply]
So this is what I said above, you are using `nvm` and systemd most probably does not use initialise `nvm`, thus launching node using your system package.

I would recommend you to uninstall Node and `nvm` altogether, and install Node 18 the same way we use in the official docs:

```
curl -sL https://deb.nodesource.com/setup_18.x | bash -
```

(further documentation here: https://github.com/nodesource/distributions)

**Reply by charismeki** [reply]
Ok. Thank you very much for your hints. I need to keep the node.js version running at the moment but will start changing that by end of June. I will be back then.

**Reply by renchap** [reply]
An alternative would be to have systemd use nvm.

Here is an example: https://gist.github.com/patoi/f725a9a39d0145bcda4c3796b6419db7

I did not test this, but you get the idea.

**Comment by charismeki** [comment]
Hi there, I purged node and npm. Then edited /etc/systemd/system/mastodon-streaming.service
and changed 
ExecStart=/usr/bin/node ./streaming
to
ExecStart=/usr/local/bin/node ./streaming.
Running now.
Thx and best,
c--

**Comment by michaelchadwick** [comment]
Thank you for helping me figure out that I had multiple versions of node installed, but Mastodon was using the system one that was, uh, a bit old (v12). All fixed now!
